File: lab01/builder/src/main/java/com/example/builder/enums/CharacterType.java
package com.example.builder.enums;

import java.util.function.Supplier;

import com.example.builder.builder.ArcherBuilder;
import com.example.builder.builder.CharacterBuilder;
import com.example.builder.builder.KnightBuilder;
import com.example.builder.builder.MageBuilder;

public enum CharacterType {
    KNIGHT(KnightBuilder::new),
    MAGE(MageBuilder::new),
    ARCHER(ArcherBuilder::new);

    private final Supplier<CharacterBuilder> builder;

    CharacterType(Supplier<CharacterBuilder> builder) {
        this.builder = builder;
    }

    public CharacterBuilder createBuilder() {
        return builder.get();
    }
}

File: lab01/builder1/Enums/CharacterType.cs
using Builder1.Builders;

namespace Builder1.Enums;

public enum CharacterType
{
    Knight,
    Mage,
    Archer,
}

public static class CharacterTypeExtensions
{
    public static ICharacterBuilder CreateBuilder(this CharacterType type) => type switch
    {
        CharacterType.Knight => new KnightBuilder(),
        CharacterType.Mage => new MageBuilder(),
        CharacterType.Archer => new ArcherBuilder(),
        _ => throw new ArgumentOutOfRangeException(nameof(type), type, null)
    };
}
